src/Components/Components/src/NavigationManager.cs

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Components.Routing;
+
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// Provides an abstraction for querying and mananging URI navigation.
+    /// </summary>
+    public abstract class NavigationManager
+    {
+        /// <summary>
+        /// An event that fires when the navigation location has changed.
+        /// </summary>
+        public event EventHandler<LocationChangedEventArgs> LocationChanged
+        {
+            add
+            {
+                AssertInitialized();
+                _locationChanged += value;
+            }
+            remove
+            {
+                AssertInitialized();
+                _locationChanged -= value;
+            }
+        }
+
+        private EventHandler<LocationChangedEventArgs> _locationChanged;
+
+        // For the baseUri it's worth storing both the string form and Uri form and
+        // keeping them in sync. These are always represented as absolute URIs with
+        // a trailing slash.
+        private Uri _baseUri;
+        private string _baseUriString;
+
+        // The URI. Always represented an absolute URI.
```

Comment:
We don't do anything to assert or verify this, do we?

Reply:
Correct for now.